src/dodal/devices/apple2_undulator.py

```diff
@@ -0,0 +1,232 @@
+
+    """
+
+    def __init__(
+        self,
+        prefix: str,
+        top_outer: str,
+        top_inner: str,
+        btm_inner: str,
+        btm_outer: str,
+        name: str = "",
+    ):
+        # Gap demand set point and readback
+        with self.add_children_as_readables():
+            self.top_outer = UndulatorPhaseMotor(prefix=prefix, infix=top_outer)
+            self.top_inner = UndulatorPhaseMotor(prefix=prefix, infix=top_inner)
+            self.btm_inner = UndulatorPhaseMotor(prefix=prefix, infix=btm_inner)
+            self.btm_outer = UndulatorPhaseMotor(prefix=prefix, infix=btm_outer)
+        # Nothing move until this is set to 1 and it will return to 0 when done
+        self.set_move = epics_signal_rw(int, prefix + "BLGSETP")
+        self.gate = epics_signal_r(UndulatorGatestatus, prefix + "BLGATE")
+        split_pv = prefix.split("-")
+        self.fault = epics_signal_r(
+            float,
+            split_pv[0] + "-" + split_pv[1] + "-STAT" + "-" + split_pv[3] + "ANYFAULT",
+        )
+        super().__init__(name=name)
+
+    @AsyncStatus.wrap
+    async def set(self, value: Apple2PhasesVal) -> None:
+        LOGGER.info(f"Setting {self.name} to {value}")
+
+        if await self.fault.get_value() != 0:
+            raise RuntimeError(f"{self.name} is in fault state")
+        if await self.gate.get_value() == UndulatorGatestatus.open:
+            raise RuntimeError(f"{self.name} is already in motion.")
+
+        await asyncio.gather(
+            self.top_outer.user_setpoint.set(value=value.top_outer),
+            self.top_inner.user_setpoint.set(value=value.top_inner),
+            self.btm_inner.user_setpoint.set(value=value.btm_inner),
+            self.btm_outer.user_setpoint.set(value=value.btm_outer),
+        )
+        timeout = await self._cal_timeout()
+        await self.set_move.set(value=1)
+        await wait_for_value(self.gate, UndulatorGatestatus.close, timeout=timeout)
+
+    async def _cal_timeout(self) -> float:
+        velos = await asyncio.gather(
+            self.top_outer.velocity.get_value(),
+            self.top_inner.velocity.get_value(),
+            self.btm_inner.velocity.get_value(),
+            self.btm_outer.velocity.get_value(),
+        )
+        cur_pos = await asyncio.gather(
+            self.top_outer.user_setpoint_demand_readback.get_value(),
+            self.top_inner.user_setpoint_demand_readback.get_value(),
+            self.btm_inner.user_setpoint_demand_readback.get_value(),
+            self.btm_outer.user_setpoint_demand_readback.get_value(),
+        )
+        target_pos = await asyncio.gather(
+            self.top_outer.user_setpoint_readback.get_value(),
+            self.top_inner.user_setpoint_readback.get_value(),
+            self.btm_inner.user_setpoint_readback.get_value(),
+            self.btm_outer.user_setpoint_readback.get_value(),
+        )
+        move_time = np.abs(np.divide(tuple(np.subtract(target_pos, cur_pos)), velos))
```

Comment:
what does it do? can we split this long line? tuple of a difference in position  as one thing, divided by velocity as a second thing. also I assume move_time is not a scalar float, we could add a type hint to it that it's np.ndarray.

by default I'd expect a variable named `move_time` to *be* a scalar float